### ISSUE TITLE
Allow unique key constraints on non-partition keys

### ISSUE THREAD

**Issue by lithp** [citusdata/citus#720]
(Maybe this shouldn't be built into Citus, and instead should be something we document.)

I think with Citus 9.2 (single-shard transactions) it's now possible to support non-partition unique columns.  You might be able to maintain a local table on the master with (part_col, unique_col UNIQUE), and use transactions to update it with the remote table.

**Comment by marcocitus** [comment]
This gets a little tricky when doing deletes, updates, insert/select, upsert, since the master wouldn't know the value being written. I suppose we could do add returning to all queries and process the results and roll back if there's a violation. It gets really tricky in MX.

**Comment by marcocitus** [comment]
Actually, it might not be so tricky in MX. We could have 2 distributed tables, table A distributed by part_col and table B distribute by unique_col with a unique constraint on unique_col. Since the workers have the distributed tables, we could put a trigger on the shards of table A that updates table B and errors out if there is a constraint violation.

**Comment by lithp** [comment]
Yeah, that would work. It has the same problems as above, where you have to interpret how deletes and updates might change the set of unique_col values. It'll also take some effort to get playing nicely with the new connection management. If you do an insert and keep your txn open, the node also needs to hold an open txn to the table B shard. w/o 2PC, that can lead to inconsistencies under node failures.
